ddtrace/debugging/_origin/span.py

```diff
@@ -0,0 +1,248 @@
+            func_qname=function,
+            evaluate_at=ProbeEvaluateTimingForMethod.ENTER,
+            template=message,
+            segments=[LiteralTemplateSegment(message)],
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+
+@attr.s
+class ExitSpanProbe(LogLineProbe):
+    __span_class__ = "exit"
+
+    @classmethod
+    def build(cls, name: str, filename: str, line: int) -> "ExitSpanProbe":
+        message = f"{cls.__span_class__} span info for {name}, in {filename}, at {line}"
+
+        return cls(
+            probe_id=str(uuid.uuid4()),
+            version=0,
+            tags={},
+            source_file=filename,
+            line=line,
+            template=message,
+            segments=[LiteralTemplateSegment(message)],
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+    @classmethod
+    def from_frame(cls, frame: FrameType) -> "ExitSpanProbe":
+        code = frame.f_code
+        return t.cast(
+            ExitSpanProbe,
+            cls.build(
+                name=code.co_qualname if sys.version_info >= (3, 11) else code.co_name,  # type: ignore[attr-defined]
+                filename=str(Path(code.co_filename).resolve()),
+                line=frame.f_lineno,
+            ),
+        )
+
+
+@dataclass
+class EntrySpanLocation:
+    name: str
+    start_line: int
+    end_line: int
+    file: str
+    module: str
+    probe: EntrySpanProbe
+
+
+class EntrySpanWrappingContext(WrappingContext):
+    def __init__(self, f):
+        super().__init__(f)
+
+        lines = linenos(f)
+        start_line = min(lines)
+        filename = str(Path(f.__code__.co_filename).resolve())
+        name = f.__qualname__
+        module = f.__module__
+        self.location = EntrySpanLocation(
+            name=name,
+            start_line=start_line,
+            end_line=max(lines),
+            file=filename,
+            module=module,
+            probe=t.cast(EntrySpanProbe, EntrySpanProbe.build(name=name, module=module, function=name)),
+        )
+
+    def __enter__(self):
+        super().__enter__()
+
+        root = ddtrace.tracer.current_root_span()
+        location = self.location
+        if root is None or root.get_tag("_dd.entry_location.file") is not None:
+            return
+
+        # Add tags to the local root
+        root.set_tag_str("_dd.entry_location.file", location.file)
+        root.set_tag_str("_dd.entry_location.start_line", str(location.start_line))
+        root.set_tag_str("_dd.entry_location.end_line", str(location.end_line))
+        root.set_tag_str("_dd.entry_location.type", location.module)
+        root.set_tag_str("_dd.entry_location.method", location.name)
+
+        if config._trace_span_origin_enriched:
```

Comment:
why we take snapshot at entry and not on exit? I think it better to snapshot exit with locals and return values?
do we capture locals at the moment or just return here?

Reply:
second, I think we also need to look at _dd.p.debug flag for the enriched snapshots.